File: medium/0622_design_circular_queue/solution.py
class MyCircularQueue:
    def __init__(self, k: int):
        self.buffer = [0] * k

        self.size = 0
        self.capacity = k

        self.front = 0
        self.back = k - 1

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False

        self.back = (self.back + 1) % self.capacity
        self.buffer[self.back] = value
        self.size += 1

        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False

        self.front = (self.front + 1) % self.capacity
        self.size -= 1

        return True

    def Front(self) -> int:
        return self.buffer[self.front] if not self.isEmpty() else -1

    def Rear(self) -> int:
        return self.buffer[self.back] if not self.isEmpty() else -1

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity

File: medium/0622_design_circular_queue/solution.cpp
#include <vector>

class MyCircularQueue {
public:
  MyCircularQueue(int k) {
    capacity = static_cast<size_t>(k);
    size = 0;

    buffer = std::vector<int>(capacity);

    front = 0;
    back = capacity - 1;
  }

  bool enQueue(int value) {
    if (isFull())
      return false;

    back = (back + 1) % capacity;
    buffer[back] = value;
    ++size;

    return true;
  }

  bool deQueue() {
    if (isEmpty())
      return false;

    front = (front + 1) % capacity;
    --size;

    return true;
  }

  int Front() { return !isEmpty() ? buffer[front] : -1; }

  int Rear() { return !isEmpty() ? buffer[back] : -1; }

  bool isEmpty() { return !static_cast<bool>(size); }

  bool isFull() { return size == capacity; }

private:
  size_t capacity;
  size_t size;

  std::vector<int> buffer;

  size_t front;
  size_t back;
};
